### PR TITLE
WIP: Removed format specific Timesteps

### LINKED ISSUE #1575: `ts.dimensions` object with different trajectory formats
### Expected behaviour

The `ts.dimensions` returned object (the numpy array of values) is NOT linked back to the state of the trajectory -- the property should return a numpy array of dimensions that has no connection to the current state of trajectory iteration once the `value` in`value = ts.dimensions` has been stored i.e., in a list.

Futhermore, this behavior must be consistent between different trajectory formats. 

### Actual behaviour

I'm so confused because look at this behavior when tracking dimensions over time with two different formats (and for even more confusion, look at what happens to the stored values for the XTC format after iteration ends!!!)

```python
  1 import MDAnalysis                                                               
  2 from MDAnalysisTests.datafiles import GRO, XTC, PSF, DCD                        
  3                                                                                 
  4 u1 = MDAnalysis.Universe(GRO, XTC)                                              
  5 u2 = MDAnalysis.Universe(PSF, DCD)                                              
  6 list_dims_1 = []                                                                
  7 list_dims_2 = []                                                                
  8                                                                                 
  9 for universe, list_dims in zip([u1, u2],                                        
 10                                [list_dims_1, list_dims_2]):                     
 11     for ts in universe.trajectory[:2]:                                          
 12         frame = ts.frame                                                        
 13         dims = ts.dimensions                                                    
 14         dims[0] = frame                                                         
 15         list_dims.append((frame, dims))                                         
 16         print('universe:', universe)                                            
 17         print('frame:', frame)                                                  
 18         print('dims:', dims)                                                    
 19         print('list_dims:', list_dims)                                          
 20                                                                                 
 21 # if the above isn't crazy enough for you                                       
 22 # when the trajectory iteration ends, the                                       
 23 # apparent XTC rewind will reset all the dimension                              
 24 # values assigned to the list to those from                                     
 25 # the first frame !!!!                                                          
 26                                                                                 
 27 for element in list_dims_1:                                                     
 28     print(element)                                                              
 29                                                                                 
 30 for element in list_dims_2:                                                     
 31     print(element) 
```

Output:
```
universe: <Universe with 47681 atoms>
frame: 0
dims: [  0.          80.01700592  80.01700592  60.          60.          90.        ]
list_dims: [(0, array([  0.        ,  80.01700592,  80.01700592,  60.        ,
        60.        ,  90.        ], dtype=float32))]
universe: <Universe with 47681 atoms>
frame: 1
dims: [  1.          80.13008118  80.13008118  60.          60.          90.        ]
list_dims: [(0, array([  1.        ,  80.13008118,  80.13008118,  60.        ,
        60.        ,  90.        ], dtype=float32)), (1, array([  1.        ,  80.13008118,  80.13008118,  60.        ,
        60.        ,  90.        ], dtype=float32))]
universe: <Universe with 3341 atoms>
frame: 0
dims: [  0.   0.   0.  90.  90.  90.]
list_dims: [(0, array([  0.,   0.,   0.,  90.,  90.,  90.], dtype=float32))]
universe: <Universe with 3341 atoms>
frame: 1
dims: [  1.   0.   0.  90.  90.  90.]
list_dims: [(0, array([  0.,   0.,   0.,  90.,  90.,  90.], dtype=float32)), (1, array([  1.,   0.,   0.,  90.,  90.,  90.], dtype=float32))]


(0, array([ 80.01700592,  80.01700592,  80.01700592,  60.        ,
        60.        ,  90.        ], dtype=float32))
(1, array([ 80.01700592,  80.01700592,  80.01700592,  60.        ,
        60.        ,  90.        ], dtype=float32))
(0, array([  0.,   0.,   0.,  90.,  90.,  90.], dtype=float32))
(1, array([  1.,   0.,   0.,  90.,  90.,  90.], dtype=float32))
```

This is incredibly confusing to me. Why on earth is the interaction behavior of the XTC returned dimension value so different from DCD returned dimension value. The latter looks sane, but XTC looks wild and has caused me hours of confusing / debugging in a much larger framework context today.

I'm using Python 3.6 and the dev branch.

### PR DESCRIPTION
**Pull request by richardjgowers**
Fixes #1575

So this is squishing all the format specific Timesteps we have.  This mostly involves moving the conversion of unitcell information from the Timestep to the Reader.

Theres's only one custom Timestep remaining for MemoryReader.  This works slightly differently (I think?) because `Timestep.positions` provides a view onto the big chunk of memory that the MemoryReader operates on.

Very much work in progress... @kain88-de thoughts?

### CODE REVIEW

File: package/MDAnalysis/coordinates/DCD.py [hunk 1]
```diff
@@ -146,7 +146,7 @@ def __init__(self, filename, convert_units=True, dt=None, **kwargs):
         self.skip_timestep = self._file.header['nsavc']
 
         self._ts_kwargs['dt'] = dt
-        self.ts = self._Timestep(self.n_atoms, **self._ts_kwargs)
+        self.ts = base.Timestep(self.n_atoms, **self._ts_kwargs)
```
Comment (kain88-de) [inline package/MDAnalysis/coordinates/DCD.py]: I didn't delete that already?

Comment (kain88-de) [reply package/MDAnalysis/coordinates/DCD.py]: Shouldn't we also delete `self._Timestep` then?

File: package/MDAnalysis/coordinates/GRO.py [hunk 2]
```diff
@@ -163,23 +140,23 @@ def _read_first_frame(self):
...
-
+            self.ts = ts = base.Timestep(n_atoms, **self._ts_kwargs)
             missed_vel = False
 
             grofile.seek(0)
             for pos, line in enumerate(grofile, start=-2):
                 # 2 header lines, 1 box line at end
                 if pos == n_atoms:
-                    unitcell = np.float32(line.split())
+                    unitcell = list(map(float, line.split()))
```
Comment (kain88-de) [inline package/MDAnalysis/coordinates/GRO.py]: I thought we benchmarked those to use `np.float32`?

Comment (orbeckst) [reply package/MDAnalysis/coordinates/GRO.py]: Yes, @kain88-de is right. We determined that the fastest conversion is letting numpy do everything.

File: package/MDAnalysis/coordinates/DLPoly.py [hunk 3]
```diff
@@ -125,17 +110,17 @@ def _read_first_frame(self):
...
+                                     forces=has_forces,
+                                     **self._ts_kwargs)
         ts._pos = coords
         if has_vels:
             ts._velocities = velocities
         if has_forces:
             ts._forces = forces
         if not imcon == 0:
-            ts._unitcell = unitcell
+            ts.triclinic_dimensions = unitcell
```
Comment (orbeckst) [inline package/MDAnalysis/coordinates/DLPoly.py]: Why add `Timestep.triclinic_dimensions` ? Is this the "original" unitcell – and if so, why not keep it as a private `_unitcell`. The user should not need to worry about box representations. We simply provide `ts.dimensions` and that's it.

File: package/MDAnalysis/coordinates/DLPoly.py [hunk 4]
```diff
@@ -159,23 +143,24 @@ def __init__(self, filename, **kwargs):
...
         if not line.startswith('timestep'):
             raise IOError
         if not self._imcon == 0:
-            ts._unitcell[0] = self._file.readline().split()
-            ts._unitcell[1] = self._file.readline().split()
-            ts._unitcell[2] = self._file.readline().split()
+            unitcell[0] = self._file.readline().split()
+            unitcell[1] = self._file.readline().split()
+            unitcell[2] = self._file.readline().split()
+            ts.triclinic_dimensions = unitcell
```
Comment (orbeckst) [inline package/MDAnalysis/coordinates/DLPoly.py]: I don't see the point of `ts.triclinic_dimensions`

File: package/MDAnalysis/coordinates/GRO.py [hunk 5]
```diff
@@ -97,47 +94,28 @@
...
 
 
-class Timestep(base.Timestep):
-    _ts_order_x = [0, 3, 4]
-    _ts_order_y = [5, 1, 6]
-    _ts_order_z = [7, 8, 2]
-
-    def _init_unitcell(self):
-        return np.zeros(9, dtype=np.float32)
+"""unitcell dimensions (A, B, C, alpha, beta, gamma)
```
Comment (orbeckst) [inline package/MDAnalysis/coordinates/GRO.py]: Should these docs be properly formatted for module level docs? From the diff I cannot really see what's going on.

File: package/MDAnalysis/coordinates/GRO.py (hunk 2 above)
Comment (orbeckst) [reply package/MDAnalysis/coordinates/GRO.py]: yes

File: package/MDAnalysis/coordinates/PDBQT.py [hunk 6]
```diff
@@ -163,9 +163,7 @@ def _read_first_frame(self):
                     # convert all entries at the end once for optimal speed
                     coords.append([line[30:38], line[38:46], line[46:54]])
         self.n_atoms = len(coords)
-        self.ts = self._Timestep.from_coordinates(
-            coords,
-            **self._ts_kwargs)
+        self.ts = base.Timestep.from_coordinates(coords, **self._ts_kwargs)
         self.ts._unitcell[:] = unitcell
```
Comment (orbeckst) [inline package/MDAnalysis/coordinates/PDBQT.py]: Didn't you want to remove `_unitcell`?

File: package/MDAnalysis/coordinates/PQR.py [hunk 7]
```diff
@@ -132,8 +132,7 @@ def _read_first_frame(self):
                     coords.append(fields[-5:-2])
 
         self.n_atoms = len(coords)
-        self.ts = self._Timestep.from_coordinates(
-            coords, **self._ts_kwargs)
+        self.ts = base.Timestep.from_coordinates(coords, **self._ts_kwargs)
         self.ts._unitcell[:] = unitcell
```
Comment (orbeckst) [inline package/MDAnalysis/coordinates/PQR.py]: Didn't you want to remove `_unitcell`?

File: package/MDAnalysis/coordinates/TRZ.py [hunk 8]
```diff
@@ -37,45 +37,6 @@
 Classes
 -------
 
-.. autoclass:: MDAnalysis.coordinates.TRZ.Timestep
-   :members:
-
-   .. attribute:: frame
```
Comment (orbeckst) [inline package/MDAnalysis/coordinates/TRZ.py]: are all of these special values now in `Timestep.data`?

File: package/MDAnalysis/coordinates/base.py [hunk 9]
```diff
@@ -274,7 +273,9 @@ def __init__(self, n_atoms, **kwargs):
         self.has_velocities = kwargs.get('velocities', False)
         self.has_forces = kwargs.get('forces', False)
 
-        self._unitcell = self._init_unitcell()
+        self.order = kwargs.get('order', 'F')
+
+        self._unitcell = np.zeros(6, dtype=np.float32)
```
Comment (orbeckst) [inline package/MDAnalysis/coordinates/base.py]: What does `_unitcell` hold now if it's not the original box?

**Comment by kain88-de** [comment]
[gist](https://gist.github.com/kain88-de/c32bcc6ae70610ae0b1cf1631b2e21f1)

I did a quick check. If we use double precision for unitcell representation then all conversions between triclinic and box are accurate up to 8 decimal places. This should be enough for us and mean we can skip keeping track of the original since the two are very very close.

**Comment by richardjgowers** [comment]
Ok so to clarify this PR, 

**Input**: `ts._unitcell` is the internal storage of the unit cell.  It is currently is the [a, b, c, alpha, beta, gamma] representation for ALL Readers, (because we've killed off custom Timestep objects).  This means that ALL Readers now have to fill `_unitcell` with information in that form.

**Access** `ts.dimensions` returns the same [a, b, c, alpha, beta, gamma] representation. `ts.triclinic_dimensions` returns the 3x3 array version of the box instead.  By having these as properties we can allow settings (eg `ts.triclinic_dimensions = ` will populate `ts._unitcell` properly.

But having reread #1575, if we want `ts.dimensions` to return a view/pointer, we can't really keep `ts.triclinic_dimensions`, as it won't return a view/pointer and will lead to confusion later on.  So I think I'm going to have to deprecate `triclinic_dimensions`.

@kain88-de I'm guessing that single precision doesn't work well enough?  So dimensions should always be float64?

**Comment by kain88-de** [comment]
> @kain88-de I'm guessing that single precision doesn't work well enough? So dimensions should always be float64?

yeah. You can replace the functions at the end of the notebook with the equivalents of the last stable release (add `mda.lib.mdamath.` at the beginning). Then the tests fail. The only change I made in comparison to the code in MDAnalysis was to remove the type definitions to single precision. 



> But having reread #1575, if we want ts.dimensions to return a view/pointer, we can't really keep ts.triclinic_dimensions, as it won't return a view/pointer and will lead to confusion later on. So I think I'm going to have to deprecate triclinic_dimensions.

We don't we allow `ts.dimensions` and `ts.tricilinic_vectors` to have a setters that stores the result in a `ts._dim` variable that gets updated by a copy with `self._dim[:] = dims`. Then we can keep the functions and we only do a copy on the assigment and reading is still a faster reference. 

It doesn't solve #1575 for which the only solution is to generate a new timestep on a read instead of updating an old one.

**Comment by richardjgowers** [comment]
@kain88-de I think what you call `ts._dim` is just what `ts._unitcell` is currently, both `ts.dimensions` and `ts.triclinic_dimensions` have setters which update `ts._unitcell`

**Comment by kain88-de** [comment]
@richardjgowers do we want to forbid setting `_unitcell` directly? We could add a setter for it that throws an exception when it's called. That way we are consistent with the usage within the library?

**Comment by richardjgowers** [comment]
@kain88-de it's a "private" attribute, if you're setting it directly then you're opening the door to weird behaviour and tears and that's not our problem :) We should probably rewrite the Readers to behave WRT that though

**Comment by kain88-de** [comment]
Yeah true. But it might help to find which reader is a bad boy that manipulates the private attribute directly.

File: package/MDAnalysis/coordinates/DCD.py (hunk 1 above)
Comment (richardjgowers) [reply package/MDAnalysis/coordinates/DCD.py]: yeah self._Timestep is a thing of the past

File: package/MDAnalysis/coordinates/GRO.py (hunk 5 above)
Comment (richardjgowers) [reply package/MDAnalysis/coordinates/GRO.py]: yeah this is an old comment from withing the format specific timestep, I kept it around as it seemd useful for developers

**Comment by richardjgowers** [comment]
Yeah I think I need to go through and make sure all unitcell info is double

**Comment by richardjgowers** [comment]
tangled mess of trying to reduce classes...

**Pull request closed without merging**